### PR TITLE
Automate puppet release

### PR DESCRIPTION
**Pull request by jcheng-splunk**
- Add github workflow to push `puppet-v<VERSION>` tag whenever the version in metadata.json is updated
- Add `puppet-release` job to gitlab CI to build and publish the module to puppet forge with blacksmith for new tags

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -55,6 +55,7 @@ compile:
     variables:
       - $CI_COMMIT_BRANCH == "main"
       - $CI_COMMIT_TAG =~ /^v[0-9]+\.[0-9]+\.[0-9]+.*/
+  retry: 2
```
Comment (dmitryax) [inline .gitlab-ci.yml:58]: looks unrelated to the PR... can you submit another one?

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -297,3 +299,25 @@ ansible-release:
...
+puppet-release:
+  image: 'docker-hub.repo.splunkdev.net/ruby:2.6-buster'
+  stage: release
+  only:
+    variables:
+      - $CI_COMMIT_TAG =~ /^puppet-v[0-9]+\.[0-9]+\.[0-9]+.*/
+  cache:
+    key: "puppet-bundler-cache"
+    paths:
+      - .cache/bundler
```
Comment (dmitryax) [inline .gitlab-ci.yml:312]: Do we really need to cache the bundler stuff assuming puppet release happens not that often? Wouldn't be better to use an up-to-date bundler installation instead?

File: deployments/puppet/README.md [hunk 3]
```diff
@@ -48,7 +48,7 @@ This class accepts the following parameters:
...
+| `manage_repo` | **Linux only**: In cases where the collector and fluentd apt/yum repositories are managed externally, set this to `false` to disable management of the repositories by this module.  **Note:** If set to `false`, the externally managed repositories should provide the `splunk-otel-collector` and `td-agent` packages.  Also, the apt (`/etc/apt/sources.list.d/splunk-otel-collector.list`, `/etc/apt/sources.list.d/splunk-td-agent.list`) and yum (`/etc/yum.repos.d/splunk-otel-collector.repo`, `/etc/yum.repos.d/splunk-td-agent.repo`) repository definition files will be deleted if they exist in order to avoid any conflicts. | `true` |
```
Comment (dmitryax) [inline deployments/puppet/README.md:51]: Same comment here. Let's move to another PR and have clean git history :)

File: deployments/puppet/Dockerfile [hunk 4]
```diff
@@ -1,31 +1,27 @@
...
+COPY ./ ./
 RUN gem install bundler && bundle install
 
 RUN mkdir -p /root/.config/puppet && \
     echo "---\n\
 disabled: true" > /root/.config/puppet/analytics.yml
 
 ENV PATH=/opt/puppetlabs/bin:$PATH
-
-COPY ./ ./
```
Comment (dmitryax) [inline deployments/puppet/Dockerfile:31]: This is unrelated to the automation, right? Can we merge it before this PR?

Comment (jcheng-splunk) [reply deployments/puppet/Dockerfile:31]: These changes are required for blacksmith and the existing integration tests.

File: .gitlab-ci.yml (hunk 1 above)
Comment (jcheng-splunk) [reply .gitlab-ci.yml]: Moved to #482

File: .gitlab-ci.yml (hunk 2 above)
Comment (jcheng-splunk) [reply .gitlab-ci.yml]: Removed bundler cache

File: deployments/puppet/README.md (hunk 3 above)
Comment (jcheng-splunk) [reply deployments/puppet/README.md]: Moved to #483

**Review by dmitryax — APPROVED** [review]
